app/src/main/java/org/mozilla/reference/browser/tabs/TabsTrayFragment.kt

```diff
@@ -74,4 +135,60 @@ class TabsTrayFragment : androidx.fragment.app.Fragment(), BackHandler {
             commit()
         }
     }
+
+    private fun invalidateActions() {
+        for (action in browserActions) {
+            val visible = action.actual.visible()
+
+            if (!visible && action.view != null) {
+                // Action should not be visible anymore. Remove view.
+                toolbar.removeView(action.view)
+                action.view = null
+            } else if (visible && action.view == null) {
+                // Action should be visible. Add view for it.
+                action.actual.createView(toolbar).let {
+                    action.view = it
+                    toolbar.addView(it)
+                }
+            }
+
+            action.view?.let { action.actual.bind(it) }
+        }
+        tabsFeature?.invalidate()
+    }
```

Comment:
The reason for the weird indirection is because I wanted the regular tabs icon to unselect when the private tabs icon was selected. The only way I found that I could do that, was by calling the `TwoStateButton#bind` which meant holding onto the view and the button together. Which was duplication from the BrowserToolbar.

Is there a better way to use the buttons to achieve this use case? 🤔 

I didn't want to add Actions in the menu so I could imitate Fennec's buttons on the left with menu options for closing all tabs on the right. Gif below:

<a href="https://giant.gfycat.com/SlowMilkyBass.webm">![slowmilkybass](https://user-images.githubusercontent.com/1370580/48960934-ceb85d80-ef3e-11e8-8bc2-323c775d5ed9.gif)</a>

Reply:
If you ignore all the indirection then you are adding views to the toolbar and updating the views. Could this be written as this directly? Either by:
* Adding views and updating them manually
* Adding a `CompoundButton` (or one of its sub classes) and style it as the button you want
* Adding a custom `TabSwitcher` that is a `ViewGroup` and deals with all the buttons and state updating internally.

Reply:
@pocmo Done!

I think I misunderstood our intentions with our UI components that we've built. I tried way too hard to make use of them. 🤦‍♂️